editor/src/components/canvas/dom-sampler.tsx

```diff
@@ -0,0 +1,472 @@
+    }
+  }
+
+  let result: { metadata: ElementInstanceMetadataMap; tree: ElementPathTrees }
+  if (elementsToCollect == 'rerender-all-elements') {
+    result = collectMetadataForPaths({
+      metadataToUpdate_MUTATE: {},
+      canvasRootContainer: canvasRootContainer,
+      pathsToCollect: validPaths,
+      validPaths: validPaths,
+      scale: options.scale,
+      selectedViews: options.selectedViews,
+      spyCollector: options.spyCollector,
+      spyPaths: spyPaths,
+    })
+  } else {
+    result = collectMetadataForPaths({
+      metadataToUpdate_MUTATE: { ...options.metadataToUpdate }, // shallow cloning this object so we can mutate it
+      canvasRootContainer: canvasRootContainer,
+      pathsToCollect: validPaths.filter((vp) =>
+        elementsToCollect.some((e) => {
+          const staticElement = EP.makeLastPartOfPathStatic(e)
+          return EP.pathsEqual(staticElement, vp) || EP.isParentOf(staticElement, vp)
+        }),
+      ),
+      validPaths: validPaths,
+      scale: options.scale,
+      selectedViews: options.selectedViews,
+      spyCollector: options.spyCollector,
+      spyPaths: spyPaths,
+    })
+  }
+
+  return result
+}
+
+function collectMetadataForPaths({
+  canvasRootContainer,
+  pathsToCollect,
+  validPaths,
+  metadataToUpdate_MUTATE,
+  scale,
+  selectedViews,
+  spyCollector,
+  spyPaths,
+}: {
+  canvasRootContainer: HTMLElement
+  pathsToCollect: Array<ElementPath>
+  validPaths: Array<ElementPath>
+  metadataToUpdate_MUTATE: ElementInstanceMetadataMap
+  scale: number
+  selectedViews: Array<ElementPath>
+  spyCollector: UiJsxCanvasContextData
+  spyPaths: Array<string>
+}): {
+  metadata: ElementInstanceMetadataMap
+  tree: ElementPathTrees
+} {
+  const containerRect = getCanvasRectangleFromElement(
+    canvasRootContainer,
+    scale,
+    'without-text-content',
+    'nearest-half',
+  )
+
+  const dynamicPathsToCollect: Array<ElementPath> = pathsToCollect
+    .flatMap((staticPath) => {
+      return spyPaths.filter((spyPath) =>
+        EP.pathsEqual(EP.makeLastPartOfPathStatic(EP.fromString(spyPath)), staticPath),
+      )
+    })
+    .map(EP.fromString)
+
+  dynamicPathsToCollect.forEach((path) => {
+    const domMetadata = collectMetadataForElementPath(
+      path,
+      validPaths,
+      selectedViews,
+      scale,
+      containerRect,
+      spyCollector,
+    )
+
+    if (domMetadata == null) {
+      // if we couldn't find any dom elements for the path, we must scan through all the spy elements to find a fallback with a potentially dynamic path
+      const spyElem = spyCollector.current.spyValues.metadata[EP.toString(path)]
+      if (spyElem != null) {
+        metadataToUpdate_MUTATE[EP.toString(path)] = {
+          ...spyElem,
+        }
+      }
+      return // we couldn't find a fallback spy element, so we bail out
+    }
+
+    const validDynamicPath = path
+    const spyMetadata = spyCollector.current.spyValues.metadata[EP.toString(validDynamicPath)]
```

Comment:
The above could be simplified by moving this and the associated bailing before the above `if (domMetadata == null)` since you're doing the same check inside that